### ISSUE TITLE
Forces are applied twice or more on CSP objects when connected as client, leading to a desynchronization

### ISSUE THREAD

**Issue by tedbarth** [FirstGearGames/FishNet#263]
Unity version: 2021.3.16f1 up to 2021.3.19f1
Fish-Networking version: 3.3.2 up to 3.4.2

**Describe the bug**
When player objects are implemented with CSP while being server authoritative, the forces applied to their RigidBody2Ds (which in turn cause position, rotation and velocity changes) are applied (feels like) twice on clients, leading to a desynchronization of client and server. This problem does not appear for Host-type clients.

This can be seen in the following video, where I first start a Host on the right instance, than a client on the left. The host player always is in sync, while the client only player rotates faster on the client than it does on the server.

[![Demo](https://img.youtube.com/vi/lZ0gXDKxd9U/0.jpg)](https://www.youtube.com/embed/lZ0gXDKxd9U)
(Video on YouTube.com)

In this video the problem is made visible by not applying the ReconcileData sent from the server. If we do apply the data instead, the problem would mostly hidden by the correction, but still be visible as a jumping rotation angle when changing input rotation direction.
If I do not call my [reconciliation method](https://github.com/tedbarth/FishNet-Bug-Doubled-Forces/blob/main/Assets/Scripts/PlayerCspController.cs#L34) `[Reconcile] private void Reconcile() { ... }` for the owner per Tick, the problem disappears, even though nothing is effectively done in that method (because commented out). Obviously this is not a solution: I want reconciliation. :-)

I first suspected `RigedBody2d.inertia` which is calculated by Unity automatically and affects the rotational reaction to an angular force, like mass does. But printing them out both per tick showed that they are the same on client and server:

```
Debug.Log($"Inertia: {_body.inertia}, Mass: {_body.mass}");
```

**To Reproduce**
1. Check out and follow instructions of this demo's README.md: https://github.com/tedbarth/FishNet-Bug-Doubled-Forces

**Comment by tedbarth** [comment]
Now I upgraded to 3.4.1 and the bug still is there. Any chance you can take a look at this in the near days?

**Comment by FirstGearGames** [comment]
I don't think this is a bug, rather a setup issue. I reworked the rigidbody
prediction and it's predicted spawn with better notes which should help
many. I'm also going to update guides with more setup details.

**Comment by tedbarth** [comment]
May I suggest to check what I configured wrongly in my example, to see what piece is missing in the documentation? ;-)
(I'm stuck here and I swallowed the documentation multiple times, also created the example different times on different ways, all the same result. I don't know how to proceed on this without your advise.)

**Comment by tedbarth** [comment]
I debugged a bit more and can see, that per tick the `TimerManager` calls `Physics2D.Simulate(tick);` only once, which is fine. Also the delta looks good.

I then just added `Physics2D.Simulate((float)TimeManager.TickDelta);` to my project, so that it simulates again in the same tick. This resulted in the same speed up effect (with same amplitude) but on both peers: Client and server player objects rotates faster. So it accumulates. My conclusion is that Unity might still run `Physics2D.Simulate(tick);` internally on the client.

To proof this I have removed your call to `Physics2D.Simulate(tick);` in `TimerManager`. If my theory was right, the client should move slower. Apparently it did. So definitively the method runs twice on the client, which I think should not be the case if the tick delta is the same, is it?
If it is ok, than I wonder if it is just about Forces not being reset by either Unity after calling the Simulate() internally or by Fishnet (if it runs before Unity's Simulate() call or because forces are swapping over to the next tick loop).

**Comment by tedbarth** [comment]
Reproducible with 3.4.2 as well.

**Comment by tedbarth** [comment]
The same symptoms appear in other recent, but unresolved reports from discord priority help: 

- https://discordapp.com/channels/424284635074134018/1057874440752484392/1057874440752484392 (In video the player moves faster on one peer, forces involved)
- https://discordapp.com/channels/424284635074134018/1057164300021669948 (Overshoot for client visible on multiple videos, forces involved)
- https://discordapp.com/channels/424284635074134018/1036178555907866734 (Different velocities between peers, object gets set back while moving)

Possibly (with a lot quotation marks around) the same thing, but to few data to know:
- https://discordapp.com/channels/424284635074134018/1071178769068478595 (Connected clients are jittery, which I also experience when syncing desynct RBs)
- https://discordapp.com/channels/424284635074134018/1059976136882405436 (Experiencing jitter using CSP)
- https://discordapp.com/channels/424284635074134018/1049904944569909248 (Camera bound to physics object moves faster with CSP)
- https://discordapp.com/channels/424284635074134018/1051823709851242537 (Physical object can't rest, Could be anything, Could also be the same problem

**Comment by FirstGearGames** [comment]
I will port the RB sample scene to 2D and test just to make sure it's not a 2D specific issue.

**Comment by tedbarth** [comment]
I also ported my project to 3D (there is a branch named "3d" now, "main" stays 2D) and see the problem there, too.

**Comment by Alzurana** [comment]
This bug is tricky, I spent the entire week prior thinking it was my mistake.
I have the same issue in my project with RigidBody2D and noticed it when I tried to use the reconcilliation method to implement rubber banding.

However, I fired up the 3D example in the fishnet package and could reproduce it by commenting out the first line (writing back transform.position to the client) in the reconcilliation method. The client indeed has forces applied at least twice, server and host games do not suffer this fate.

TL;DR:
This bug affects both, 2D and 3D and becomes visible when you stop reconciling the position, for example.

**Comment by FirstGearGames** [comment]
This doesn't sound like a bug. If you're commenting out code inside the
reconcile to force the issue it sounds like you're implementation is
incorrect.

**Comment by Alzurana** [comment]
> This doesn't sound like a bug. If you're commenting out code inside the reconcile to force the issue it sounds like you're implementation is incorrect.


The underlying bug is that CSP is not predicting correctly on a client only instance at all. Commenting out that line is simply a means to make this visible. Ofc, in a functioning implementation it's not commented out.
It's merely a way to see the severe effect that physics have on the client (as in, being applied twice for some reason)
The reconcile method hides this by resetting the client, but the underlying issue is not resolved that the client actually desyncs each frame due to doubled forces and does not predict correctly.

**Comment by tedbarth** [comment]
Exactly what @Alzurana said. By not applying the reconcile data on the client we just make it more visible. If we don't do it, the behavior (Simulated too much/many time/s) is just not that visible, but still there: It gets hidden mostly by the correction but still causes the simulations to desynchronize on every frame. This probably is the reason why it wasn't seen before - The reconciliation did correct it.

**Comment by FirstGearGames** [comment]
Unless I'm misunderstanding something, it sounds like everything is working
as intended. CSP works by replaying inputs after reconciliation and if you
aren't reconciling there will be large amounts of desynchronization. Host
doesn't reconcile because server does not need to as it's never behind.

**Comment by Alzurana** [comment]
> Unless I'm misunderstanding something,

Yes, there is a misunderstanding. When the client reconciles and then replays inputs it does it with twice the force. It's playing the physics twice as often as it should. It's hardly noticable with reconciliation but it still results in awkward jitter.

From my understanding CSP should predict movement as accuratly as possible so that reconcilliate does not need to correct the actual transform data much. But what is happening is that the client is off by a lot.

I'd ask you to just try what I proposed. When you see it you will understand. Take the 3D example, comment out that line and connect via a client to a separate hosted game. Hit the movement keys on both and compare the movement. The difference is way beyond what is acceptable prediction. The reason for this excercise is to isolate what the prediction does without concealing the results with reconcilliate.

I would've made a video and dug a bit more into this but work unfortunately called me back for the week. I'll try my best to help out when I get some free time.

**Comment by FirstGearGames** [comment]
It's not possible for values to change in a replay. More likely something
is up with your CSP logic. Today is my day off but I'd be happy to take a
look tomorrow. Make a thread in our help channel and tag me. Include your
replicate and reconcile please.

**Comment by tedbarth** [comment]
Thank you for taking a look! :+1: 

I created two more videos to compare (left is always server/host, right the client):

1. Example with active reconciliation - The problem is visible only when directions are changed. In other situations it gets overridden by correction mechanism. The corrections basically works, but the object still moves too fast so that it always desynchronizes.
https://youtu.be/W7kIYjYKVRk

2. Example without active reconciliation - The problem is clearly visible (The client moves faster)
https://youtu.be/mUgyEuGBs4E

**Comment by Alzurana** [comment]
So far no luck, the only thing that I could figure out has something to do with `Physics.Simulate` and `Physics2D.Simulate` probably being called too many times.

All tests were performed within the 3D RB prediction example scene provided with FishNet.

Removing the call to those functions from the `TimeManager` will restore the client to normal behavior on predicted RBs while the server will not simulate any physics anymore. That means the client still has those functions called somewhere. It also continues to move non networked RBs on the client when inputs are happening (as part of the replay)

Instead: Removing the call to reconcile entirely on the client lets physics advance at a normal pace again. This also removes the replay loop from execution. Ofc this is not desired as we need to reconcile.

It's simply an indication that, somehow, reconcile's replay and the `TimeManager` together seem to amount to too many calls to `Physics.Simulate` and `Physics2D.Simulate`; or objects are not reset correctly before running replays.

**Comment by tedbarth** [comment]
There is a new priority help thread from a different person that seems to experience the same symptoms:
https://discord.com/channels/424284635074134018/1080014650822901822/1080014650822901822

**Comment by tedbarth** [comment]
@FirstGearGames Discussion thread opened for everyone in discord:
https://discord.com/channels/424284635074134018/1080032627119108096

**Comment by tedbarth** [comment]
All problems solved in the discord chat: We had completely wrong assumption about how reconciliation works.
In my case (for people having similar issues) it was that I split the appliance of force off of the `[Reconcile] Reconcile()` method. That made reconciliation non-deterministic: Stepping back to a state in time would not update the forces, but still apply the old state when Simulate() is called as part of one reconciliation step.

This all (including that I expected Simulate() to be called only once per frame) originated from the assumption that per reconciliation step all NetworkBehaviour components (including my Acceleration component) is experiencing a virtual network tick. Instead all NetworkBehaviour components are handled individually.

The fixed code (3D example) can be found here:
https://github.com/tedbarth/FishNet-Bug-Doubled-Forces/blob/3d-fixed/Assets/Scripts/PlayerCspController.cs

Thanks to all participants for the help. This issue frustrated me for almost three weeks. :relieved:
